src/ir/ty.rs

```diff
@@ -311,18 +311,21 @@ impl Type {
         match self.kind {
             TypeKind::Named => {
                 let name = self.name().expect("Unnamed named type?");
-                let mut chars = name.chars();
-                let first = chars.next().unwrap();
-                let mut remaining = chars;
-
-                let valid = (first.is_alphabetic() || first == '_') &&
-                            remaining.all(|c| c.is_alphanumeric() || c == '_');
-
-                !valid
+                Self::is_invalid_name(&name)
             }
             _ => false,
         }
     }
+    pub fn is_invalid_name(name: &str) -> bool {
+        let mut chars = name.chars();
+        let first = chars.next().unwrap();
```

Comment:
The fact that `name` is non-empty holds for `Named` types because we enforce it at creation, but does not necessarily hold for all inputs we provide to this, so I'd rather keep something similar to what you had:

```rust
pub fn is_valid_identifier(name: &str) -> bool {
    let mut chars = name.chars();
    chars.next().map_or(false, char::is_alphabetic) && chars.all(|c| c == '_' || c.is_alphanumeric)
}
```

Then use `!Self::is_valid_identifier` in the `Named` case.